File: legacy/src/logic/solvers/SemiImplicitEulerSolver.java
package logic.solvers;

import model.physics.Body2D;
import model.physics.Universe2D;

public class SemiImplicitEulerSolver extends Physics2DSolver
{
	public 
	SemiImplicitEulerSolver(Universe2D u) 
	{
		super(u);
		timestep = 0.1;
		displayName = "Semi-implicit Euler";
	}
	
	@Override
	public Physics2DSolver getNewInstance(Universe2D u) {
		return new SemiImplicitEulerSolver(u);
	}
	
	@Override
	public void step()
	{
		//position loop
		for(Body2D b1 : universe.bodies)
		{
			b1.position.add(b1.velocity.times(timestep));
		}
		
		computeAllBodiesAccelerations();
		
		//velocity loop
		for(Body2D b1 : universe.bodies)
		{
			b1.velocity.add(b1.acceleration.times(timestep));
			b1.position.add(b1.velocity.times(timestep));
		}
		
		timeElapsed += timestep;
	}

}

File: src/physics/solvers/semi_implicit_euler_solver.cpp
/*
 * semi_implicit_euler_solver.cpp
 *
 *  Created on: 12 de jul de 2016
 *      Author: carlosfaruolo
 */

#include "semi_implicit_euler_solver.hpp"

#include <vector>
using std::vector;

DEFINE_CLASS_FACTORY(SemiImplicitEulerSolver, "Semi-implicit Euler");

SemiImplicitEulerSolver::SemiImplicitEulerSolver(Universe2D& u)
: AbstractPhysics2DSolver(&CLASS_FACTORY, u, 0.01)
{}

void SemiImplicitEulerSolver::step()
{
	//position loop
	foreach(Body2D*, body, vector<Body2D*>, universe.bodies)
	{
		body->position += body->velocity * timestep;
	}

	computeAccelerations();

	//velocity loop
	foreach(Body2D*, body, vector<Body2D*>, universe.bodies)
	{
		body->velocity += body->acceleration * timestep;
	}

	timeElapsed += timestep;
}
